services/brig/src/Brig/API/Public.hs

```diff
@@ -938,65 +948,77 @@ getRichInfoH :: UserId ::: UserId ::: JSON -> Handler Response
+data CreateUserResponse
+  = CreateUserResponse (Public.Cookie (ZAuth.Token ZAuth.User)) UserId Public.SelfProfile
+
+createUser :: Public.NewUserPublic -> Handler CreateUserResponse
+createUser (Public.NewUserPublic new) = do
+  for_ (Public.newUserEmail new) $ checkWhitelist . Left
+  for_ (Public.newUserPhone new) $ checkWhitelist . Right
   result <- API.createUser new !>> newUserError
   let acc = createdAccount result
-  let usr = accountUser acc
   let eac = createdEmailActivation result
   let pac = createdPhoneActivation result
   let epair = (,) <$> (activationKey <$> eac) <*> (activationCode <$> eac)
   let ppair = (,) <$> (activationKey <$> pac) <*> (activationCode <$> pac)
-  let lang = userLocale usr
+  let newUserLabel = Public.newUserLabel new
+  let newUserTeam = Public.newUserTeam new
+  let usr = accountUser acc
+  let Public.User {userLocale, userDisplayName, userId} = usr
+  let userEmail = Public.userEmail usr
+  let userPhone = Public.userPhone usr
   lift $ do
-    for_ (liftM2 (,) (userEmail usr) epair) $ \(e, p) ->
-      sendActivationEmail e (userDisplayName usr) p (Just lang) (newUserTeam new)
-    for_ (liftM2 (,) (userPhone usr) ppair) $ \(p, c) ->
-      sendActivationSms p c (Just lang)
-    for_ (liftM3 (,,) (userEmail usr) (createdUserTeam result) (newUserTeam new)) $ \(e, ct, ut) ->
-      sendWelcomeEmail e ct ut (Just lang)
+    for_ (liftM2 (,) userEmail epair) $ \(e, p) ->
+      sendActivationEmail e userDisplayName p (Just userLocale) newUserTeam
+    for_ (liftM2 (,) userPhone ppair) $ \(p, c) ->
+      sendActivationSms p c (Just userLocale)
+    for_ (liftM3 (,,) userEmail (createdUserTeam result) newUserTeam) $ \(e, ct, ut) ->
+      sendWelcomeEmail e ct ut (Just userLocale)
   cok <- case acc of
-    UserAccount _ Ephemeral -> lift $ Auth.newCookie @ZAuth.User (userId usr) SessionCookie (newUserLabel new)
-    UserAccount _ _ -> lift $ Auth.newCookie @ZAuth.User (userId usr) PersistentCookie (newUserLabel new)
-  pure $ CreateUserResponse cok (userId usr) (SelfProfile usr)
+    UserAccount _ Ephemeral -> lift $ Auth.newCookie @ZAuth.User userId Public.SessionCookie newUserLabel
+    UserAccount _ _ -> lift $ Auth.newCookie @ZAuth.User userId Public.PersistentCookie newUserLabel
+  pure $ CreateUserResponse cok userId (Public.SelfProfile usr)
   where
-    sendActivationEmail e u p l (Just (NewTeamCreator (BindingNewTeamUser (Team.BindingNewTeam t) _))) =
-      sendTeamActivationMail e u p l (fromRange $ t ^. Team.newTeamName)
-    sendActivationEmail e u p l _ =
-      sendActivationMail e u p l Nothing
-    sendWelcomeEmail :: Email -> CreateUserTeam -> NewTeamUser -> Maybe Locale -> AppIO ()
+    sendActivationEmail e u p l = \case
+      (Just (Public.NewTeamCreator (Public.BindingNewTeamUser (Public.BindingNewTeam t) _))) ->
+        sendTeamActivationMail e u p l (fromRange $ t ^. Public.newTeamName)
+      _ ->
+        sendActivationMail e u p l Nothing
```

Comment:
Doesn't this loop?  (If so, the old one does, too.)

Also, I want to have an hlint rule that disallows catch-alls like this.  If there was another `Just _` and then a `Nothing`, it would have been much harder to get confused here.

Reply:
The naming is very confusing here, so I think you saw two different names as the same? There are three different functions involved:
- sendActivationMail
- sendActivation**E**mail
- send**Team**ActivationMail

So at a first glance it doesn't seem like it's looping, but maybe I can make this less confusing.

Reply:
I tried improving the pattern matching, but I don't think it really got better:

```diff
@@ -1006,9 +1007,14 @@ createUser (Public.NewUserPublic new) = do
   pure $ CreateUserResponse cok userId (Public.SelfProfile usr)
   where
     sendActivationEmail e u p l = \case
-      (Just (Public.NewTeamCreator (Public.BindingNewTeamUser (Public.BindingNewTeam t) _))) ->
-        sendTeamActivationMail e u p l (fromRange $ t ^. Public.newTeamName)
-      _ ->
+      Just (Public.NewTeamCreator newTeamUser) -> do
+        let teamName = Public._newTeamName . coerce @_ @(Public.NewTeam ()) . Public.bnuTeam $ newTeamUser
+        sendTeamActivationMail e u p l (fromRange teamName)
+      Just _ ->
+        -- not a creater
+        sendActivationMail e u p l Nothing
+      Nothing ->
+        -- not a team user at all
         sendActivationMail e u p l Nothing
     sendWelcomeEmail :: Public.Email -> CreateUserTeam -> Public.NewTeamUser -> Maybe Public.Locale -> AppIO ()
     -- NOTE: Welcome e-mails for the team creator are not dealt by brig anymore
```

So I think I'll just merge this instead of going through another CI cycle.

Reply:
This is another possibility:

```diff
@@ -1005,10 +1005,12 @@ createUser (Public.NewUserPublic new) = do
     UserAccount _ _ -> lift $ Auth.newCookie @ZAuth.User userId Public.PersistentCookie newUserLabel
   pure $ CreateUserResponse cok userId (Public.SelfProfile usr)
   where
-    sendActivationEmail e u p l = \case
-      (Just (Public.NewTeamCreator (Public.BindingNewTeamUser (Public.BindingNewTeam t) _))) ->
-        sendTeamActivationMail e u p l (fromRange $ t ^. Public.newTeamName)
-      _ ->
+    sendActivationEmail e u p l mTeamUser
+      | Just teamUser <- mTeamUser,
+        Public.NewTeamCreator creator <- teamUser,
+        let Public.BindingNewTeamUser (Public.BindingNewTeam team) _ = creator =
+        sendTeamActivationMail e u p l (fromRange $ team ^. Public.newTeamName)
+      | otherwise =
         sendActivationMail e u p l Nothing
     sendWelcomeEmail :: Public.Email -> CreateUserTeam -> Public.NewTeamUser -> Maybe Public.Locale -> AppIO ()
     -- NOTE: Welcome e-mails for the team creator are not dealt by brig anymore
```

Or I just add a comment like "not only a `NewTeamUser`, but `NewTeamCreator`"?

I could do either of those in #1116, not sure which. Opinions?